### DISCUSSION TITLE
Cold-Warm-Hot Reset buttons

### DISCUSSION THREAD

**Discussion by moixacfz** [semuconsulting/PyGPSClient#77] [General]
Hey:)
This is a great tool. However, for my use I need the option to Cold Restart the device. It would be helpful if there would be an option to do that.
<img width="58" alt="image" src="https://github.com/semuconsulting/PyGPSClient/assets/51441089/1dbbcc0b-d49e-4efc-94dc-0a51cfee131a">
P.S. Why was this written in tkinter and not in pyQt?

**Comment by semuadmin** [comment]
Hi @moixacfz 

Glad the tool is proving useful.

To take your first query, I do in fact provide Hot, Warm and Cold Reset functions in the  [sample pygpsclient.json configuration file](https://github.com/semuconsulting/PyGPSClient/blob/master/pygpsclient.json). If you place this file in your home directory, PyGPSClient will automatically load the provided user preset UBX configuration commands and you can select them from the UBX Configuration Preset panel:

<img width="362" alt="preset cold start" src="https://github.com/semuconsulting/PyGPSClient/assets/28569967/0d4dc7d6-9adf-46e9-b5f7-c635b61c9ca4">


The relevant preset command strings are:

```
"Force HOT Reset (!!! Will require reconnection !!!), CFG, CFG-RST, 00000000, 1",
"Force WARM Reset (!!! Will require reconnection !!!), CFG, CFG-RST, 00010000, 1",
"Force COLD Reset (!!! Will require reconnection !!!), CFG, CFG-RST, ffff0000, 1",
```

These use a reset mode (byte 3 of the payload) of 0x00 = "Hardware reset (watchdog) immediately". These will typically result in an `Error in stream read failed: [ErrNo6] Device not configured` error (_which passes for an acknowledgement_), but you can simply reconnect via the PyGPSClient settings panel.

Refer to your devices Interface Specification for details of all the options available with the UBX-CFG-RST command.


In answer to your second point, it's coded in tkinter simply because that's included as standard in most Python distributions, it's well documented, it's (relatively) easy for novice contributors to learn,  it readily supports all the functionality I need, and apps can be distributed under a BSD-2 license. Admittedly, the UI looks fairly 'old school', but I'm far more concerned with functionality than cosmetics. _Plus - and this is purely a personal view - I passionately dislike the C++ Qt SDK_.

The pros and cons of tkinter vs PyQT are well documented e.g. [tkinter vs PyQt](https://www.pythonguis.com/faq/pyqt-vs-tkinter/). If I'd known it was going to evolve as it has, I might have chosen a different path, but a rewrite at this point would take more time and effort than I'm willing or able to expend. If I ever did rewrite, it would probably be as a web app using e.g. Flask (or one of the newer web frameworks) with a JavaScript UI.

**Reply by kneave** [reply]
Is there a trick to getting these presets to appear please? I have the sample file in my home folder, and have also tried to load it manually from the file -> load configuration option, but still only see the default and not any new options like cold start. I'm connecting to a F9P and everything else seems to be working, I just don't have this option and it'd be really useful for testing.

**Reply by semuadmin** [reply]
Hi @kneave 

Make sure your *.json configuration file...
- is valid (_correctly formated json_)
- is in the correct location
- is being successfully loaded on startup

By default, PyGPSClient looks for the file `pygpsclient.json` in the user's home directory. You can override this using the `--config` or `-C` command line argument e.g.

```shell
pygpsclient -C /user/myuser/myconfigfile.json
```

On startup, PyGPSClient should display something like the following message in the status bar:

<img width="321" alt="Screenshot 2025-07-08 at 04 54 15" src="https://github.com/user-attachments/assets/8a3b67fb-25d7-41bc-bee1-efd05df86e31" />

If the file is invalid (e.g. badly formatted json, or an out-of-date version), PyGPSClient will display an error message and revert to using the internal default values (which omit any user presets) e.g.

<img width="917" alt="Screenshot 2025-07-08 at 04 49 12" src="https://github.com/user-attachments/assets/2ecd4ed9-d83c-41de-9358-231d8e3a15ea" />

---
**NOTE!** there was a typo in an earlier version of the example [pygpsclient.json](https://github.com/semuconsulting/PyGPSClient/blob/master/pygpsclient.json) file which has since been rectified - try re-downloading the very latest version and ensure you are using the latest version of PyGPSClient (v1.5.10 at time of writing).

**TIP!** The safest way to ensure you have a valid, up-to-date configuration file is to use the `Save Configuration` option from the File menu, and then edit this saved file as required.

**TIP!** If you're manually editing the file, it helps to use a text editor which can validate json format as you type. The [json extension to the Visual Studio Code editor](https://github.com/ZainChen/vscode-json) is one example, and several free online editors are available.

Hope this helps.

**Reply by kneave** [reply]
The new file did the trick, thank you for the quick, detailed response, and all the rest of the hard work on this project too!
